cmd/dmverity-vhd/main.go

```diff
@@ -81,142 +85,306 @@ func main() {
-		// so it is slower but much more memory efficient
-		var opts []daemon.Option
-		if !ctx.GlobalBool(bufferedReaderFlag) {
-			opt := daemon.WithUnbufferedOpener()
-			opts = append(opts, opt)
+
+		layerIDSplit := strings.Split(layerID, "/")
+		layerDigests[layerNumber] = layerIDSplit[len(layerIDSplit)-1]
+	}
+	return layerDigests, nil
+}
+
+func isTar(reader io.Reader) (io.Reader, bool) {
+
+	// Wraps reader in :
+	//   A TeeReader which copies read bytes into a separate buffer.
+	//   A TarReader to read the header of the tar file.
+	var header bytes.Buffer
+	teeReader := io.TeeReader(reader, &header)
+	tarReader := tar.NewReader(teeReader)
+
+	_, err := tarReader.Next()
+
+	if err == nil {
+		return io.MultiReader(&header, reader), true
+	}
+
+	return io.MultiReader(&header, reader), false
+}
+
+func processLocalImage(imageReader io.Reader, onLayer LayerProcessor) (layerDigests map[int]string, layerIDs map[int]string, err error) {
+
+	imageFileReader := tar.NewReader(imageReader)
+	layerIDs = make(map[int]string)
+	layerDigestCandidates := make(map[string]map[int]string)
+	var configPath string
+	for {
+		hdr, err := imageFileReader.Next()
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			return nil, nil, err
 		}
 
-		img, err = daemon.Image(ref, opts...)
-	} else {
-		var remoteOpts []remote.Option
-		if ctx.IsSet(usernameFlag) && ctx.IsSet(passwordFlag) {
-			auth := authn.Basic{
-				Username: ctx.String(usernameFlag),
-				Password: ctx.String(passwordFlag),
+		// If the file is a tar, assume it's a layer, and call the callback
+		imageFileReader, isTar := isTar(imageFileReader)
+		if isTar {
+			if err := onLayer(hdr.Name, imageFileReader); err != nil {
+				return nil, nil, err
+			}
+		} else if hdr.Name == "manifest.json" {
+
+			type Manifest []struct {
+				Config string   `json:"Config"`
+				Layers []string `json:"Layers"`
+			}
+			var manifest Manifest
+
+			manifestData, err := io.ReadAll(imageFileReader)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			if err := json.Unmarshal(manifestData, &manifest); err != nil {
+				return nil, nil, err
+			}
+
+			configPath = manifest[0].Config
+
+			for layerNumber, layerID := range manifest[0].Layers {
+				layerIDSplit := strings.Split(layerID, ":")
+				layerIDs[layerNumber] = layerIDSplit[len(layerIDSplit)-1]
 			}
-			authConf, err := auth.Authorization()
+
+		} else { // Attempt to parse as a config file
+
+			configData, err := io.ReadAll(imageFileReader)
 			if err != nil {
-				return nil, fmt.Errorf("failed to set remote: %w", err)
+				return nil, nil, err
+			}
+
+			// Attempt to parse as if it's an image config file, trying each version until one works
+			parsingFunctions := []func([]byte) (map[int]string, error){
+				getLayerDigestsV25,
+				getLayerDigestsV24,
+			}
+			for _, parseFunc := range parsingFunctions {
+				layerDigests, err := parseFunc(configData)
```

Comment:
`layerDigests` is currently shadowing the return `layerDigests`. Maybe rename it to `ldCandidate` or something similar.

Reply:
Done in https://github.com/microsoft/hcsshim/pull/2089/commits/2b4733209d6ae6c458f9eb3c840baad152eadc6f